<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Michael Hahsler">

<title>List of Seriation Criteria implemented in the R package seriation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="seriation_criteria_files/libs/clipboard/clipboard.min.js"></script>
<script src="seriation_criteria_files/libs/quarto-html/quarto.js"></script>
<script src="seriation_criteria_files/libs/quarto-html/popper.min.js"></script>
<script src="seriation_criteria_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="seriation_criteria_files/libs/quarto-html/anchor.min.js"></script>
<link href="seriation_criteria_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="seriation_criteria_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="seriation_criteria_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="seriation_criteria_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="seriation_criteria_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#criteria-for-distances" id="toc-criteria-for-distances" class="nav-link active" data-scroll-target="#criteria-for-distances">Criteria for distances</a>
  <ul class="collapse">
  <li><a href="#sum" id="toc-sum" class="nav-link" data-scroll-target="#sum">2SUM</a></li>
  <li><a href="#ar_deviations" id="toc-ar_deviations" class="nav-link" data-scroll-target="#ar_deviations">AR_deviations</a></li>
  <li><a href="#ar_events" id="toc-ar_events" class="nav-link" data-scroll-target="#ar_events">AR_events</a></li>
  <li><a href="#bar" id="toc-bar" class="nav-link" data-scroll-target="#bar">BAR</a></li>
  <li><a href="#gradient_raw" id="toc-gradient_raw" class="nav-link" data-scroll-target="#gradient_raw">Gradient_raw</a></li>
  <li><a href="#gradient_weighted" id="toc-gradient_weighted" class="nav-link" data-scroll-target="#gradient_weighted">Gradient_weighted</a></li>
  <li><a href="#inertia" id="toc-inertia" class="nav-link" data-scroll-target="#inertia">Inertia</a></li>
  <li><a href="#lazy_path_length" id="toc-lazy_path_length" class="nav-link" data-scroll-target="#lazy_path_length">Lazy_path_length</a></li>
  <li><a href="#least_squares" id="toc-least_squares" class="nav-link" data-scroll-target="#least_squares">Least_squares</a></li>
  <li><a href="#ls" id="toc-ls" class="nav-link" data-scroll-target="#ls">LS</a></li>
  <li><a href="#mds_stress" id="toc-mds_stress" class="nav-link" data-scroll-target="#mds_stress">MDS_stress</a></li>
  <li><a href="#me" id="toc-me" class="nav-link" data-scroll-target="#me">ME</a></li>
  <li><a href="#moore_stress" id="toc-moore_stress" class="nav-link" data-scroll-target="#moore_stress">Moore_stress</a></li>
  <li><a href="#neumann_stress" id="toc-neumann_stress" class="nav-link" data-scroll-target="#neumann_stress">Neumann_stress</a></li>
  <li><a href="#path_length" id="toc-path_length" class="nav-link" data-scroll-target="#path_length">Path_length</a></li>
  <li><a href="#rgar" id="toc-rgar" class="nav-link" data-scroll-target="#rgar">RGAR</a></li>
  <li><a href="#rho" id="toc-rho" class="nav-link" data-scroll-target="#rho">Rho</a></li>
  </ul></li>
  <li><a href="#criteria-for-data-matrices-tables-and-data.frames" id="toc-criteria-for-data-matrices-tables-and-data.frames" class="nav-link" data-scroll-target="#criteria-for-data-matrices-tables-and-data.frames">Criteria for data matrices, tables and data.frames</a>
  <ul class="collapse">
  <li><a href="#cor_r" id="toc-cor_r" class="nav-link" data-scroll-target="#cor_r">Cor_R</a></li>
  <li><a href="#me-1" id="toc-me-1" class="nav-link" data-scroll-target="#me-1">ME</a></li>
  <li><a href="#moore_stress-1" id="toc-moore_stress-1" class="nav-link" data-scroll-target="#moore_stress-1">Moore_stress</a></li>
  <li><a href="#neumann_stress-1" id="toc-neumann_stress-1" class="nav-link" data-scroll-target="#neumann_stress-1">Neumann_stress</a></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">List of Seriation Criteria implemented in the R package seriation</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Michael Hahsler </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<p>This document contains the seriation criteria for judging the quality of permutations given data implemented in the R package <a href="https://github.com/mhahsler/seriation"><code>seriation</code></a>.</p>
<p>This list was created for the following version of seriation:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"seriation"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">packageVersion</span>(<span class="st">'seriation'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] '1.5.5'</code></pre>
</div>
</div>
<p>The criteria are organized by methods that evaluate the permutation based on distance data or data matrices.</p>
<p><code>merit</code> specified if the criteria function increases with better fit or if it is formulated as a loss criteria functions (<code>merit = FALSE</code>).</p>
<p>If a seriation method directly tries to optimize the criterion, than its name is specified under “optimized by”. The names can be used as the <code>method</code> argument in <code>seriate()</code>.</p>
<section id="criteria-for-distances" class="level1">
<h1>Criteria for distances</h1>
<section id="sum" class="level2">
<h2 class="anchored" data-anchor-id="sum">2SUM</h2>
<p>2-Sum Criterion: The 2-Sum loss criterion multiplies the similarity between objects with the squared rank differences (Barnard, Pothen and Simon, 1993).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: QAP_2SUM, Spectral, Spectral_norm</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="ar_deviations" class="level2">
<h2 class="anchored" data-anchor-id="ar_deviations">AR_deviations</h2>
<p>Anti-Robinson deviations: The number of violations of the anti-Robinson form weighted by the deviation (Chen, 2002).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: N/A</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="ar_events" class="level2">
<h2 class="anchored" data-anchor-id="ar_events">AR_events</h2>
<p>Anti-Robinson events: The number of violations of the anti-Robinson form (Chen, 2002).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: N/A</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="bar" class="level2">
<h2 class="anchored" data-anchor-id="bar">BAR</h2>
<p>Banded Anti-Robinson form criterion: Measure for closeness to the anti-Robinson form in a band of size b (Earle and Hurley, 2015).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: QAP_BAR</li>
</ul>
<p>additional parameters:</p>
<table class="table">
<thead>
<tr class="header">
<th></th>
<th style="text-align: left;">default values</th>
<th style="text-align: left;">help</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>b</td>
<td style="text-align: left;">NULL</td>
<td style="text-align: left;">band size defaults to a band of 20% of n</td>
</tr>
</tbody>
</table>
</section>
<section id="gradient_raw" class="level2">
<h2 class="anchored" data-anchor-id="gradient_raw">Gradient_raw</h2>
<p>Gradient measure: Evaluates how well distances increase when moving away from the diagonal of the distance matrix (Hubert et al, 2001).</p>
<ul>
<li>merit: TRUE</li>
<li>optimized by: BBURCG</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="gradient_weighted" class="level2">
<h2 class="anchored" data-anchor-id="gradient_weighted">Gradient_weighted</h2>
<p>Gradient measure (weighted): Evaluates how well distances increase when moving away from the diagonal of the distance matrix (Hubert et al, 2001).</p>
<ul>
<li>merit: TRUE</li>
<li>optimized by: BBWRCG</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="inertia" class="level2">
<h2 class="anchored" data-anchor-id="inertia">Inertia</h2>
<p>Inertia criterion: Measures the moment of the inertia of dissimilarity values around the diagonal of the distance matrix (Caraux and Pinloche, 2005).</p>
<ul>
<li>merit: TRUE</li>
<li>optimized by: QAP_Inertia</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="lazy_path_length" class="level2">
<h2 class="anchored" data-anchor-id="lazy_path_length">Lazy_path_length</h2>
<p>Lazy path length: A weighted version of the Hamiltonian path criterion where later distances are less important (Earl and Hurley, 2015).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: N/A</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="least_squares" class="level2">
<h2 class="anchored" data-anchor-id="least_squares">Least_squares</h2>
<p>Least squares criterion: The sum of squared differences between distances and the rank differences (Caraux and Pinloche, 2005).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: N/A</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="ls" class="level2">
<h2 class="anchored" data-anchor-id="ls">LS</h2>
<p>Linear Seriation Criterion: Weights the distances with the absolute rank differences (Hubert and Schultz, 1976).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: ARSA, QAP_LS</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="mds_stress" class="level2">
<h2 class="anchored" data-anchor-id="mds_stress">MDS_stress</h2>
<p>Normalized stress of a configuration given by a seriation order</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: MDS, isoMDS, Sammon_mapping, monoMDS, metaMDS</li>
</ul>
<p>additional parameters:</p>
<table class="table">
<thead>
<tr class="header">
<th></th>
<th style="text-align: left;">default values</th>
<th style="text-align: left;">help</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>warn</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">produce a warning if the 1D MDS fit does not preserve the given order (see ? seriation::uniscale).</td>
</tr>
</tbody>
</table>
</section>
<section id="me" class="level2">
<h2 class="anchored" data-anchor-id="me">ME</h2>
<p>Measure of effectiveness applied to the reordered similarity matrix (McCormick, 1972).</p>
<ul>
<li>merit: TRUE</li>
<li>optimized by: BEA, BEA_TSP</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="moore_stress" class="level2">
<h2 class="anchored" data-anchor-id="moore_stress">Moore_stress</h2>
<p>Stress criterion (Moore neighborhood) applied to the reordered similarity matrix (Niermann, 2005).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: N/A</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="neumann_stress" class="level2">
<h2 class="anchored" data-anchor-id="neumann_stress">Neumann_stress</h2>
<p>Stress criterion (Neumann neighborhood) applied to the reordered similarity matrix (Niermann, 2005).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: N/A</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="path_length" class="level2">
<h2 class="anchored" data-anchor-id="path_length">Path_length</h2>
<p>Hamiltonian path length: Sum of distances by following the permutation (Caraux and Pinloche, 2005).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: TSP, GW, GW_single, GW_average, GW_complete, GW_ward, OLO, OLO_single, OLO_average, OLO_complete, OLO_ward</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="rgar" class="level2">
<h2 class="anchored" data-anchor-id="rgar">RGAR</h2>
<p>Relative generalized anti-Robinson events: Counts Anti-Robinson events in a variable band of size w around the main diagonal and normalizes by the maximum of possible events (Tien et al, 2008).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: N/A</li>
</ul>
<p>additional parameters:</p>
<table class="table">
<thead>
<tr class="header">
<th></th>
<th style="text-align: left;">default values</th>
<th style="text-align: left;">help</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>w</td>
<td style="text-align: left;">NULL</td>
<td style="text-align: left;">window size. Default is to use a pct of 100% of n</td>
</tr>
<tr class="even">
<td>pct</td>
<td style="text-align: left;">100</td>
<td style="text-align: left;">specify w as a percentage of n in (0,100]</td>
</tr>
<tr class="odd">
<td>relative</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: left;">set to FALSE to get the GAR, i.e., the absolute number of AR events in the window.</td>
</tr>
</tbody>
</table>
</section>
<section id="rho" class="level2">
<h2 class="anchored" data-anchor-id="rho">Rho</h2>
<p>Absolute value of the Spearman rank correlation between original distances and rank differences of the order.</p>
<ul>
<li>merit: TRUE</li>
<li>optimized by: N/A</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
</section>
<section id="criteria-for-data-matrices-tables-and-data.frames" class="level1">
<h1>Criteria for data matrices, tables and data.frames</h1>
<section id="cor_r" class="level2">
<h2 class="anchored" data-anchor-id="cor_r">Cor_R</h2>
<p>Weighted correlation coefficient R: A measure of effectiveness normalized between -1 and 1 (Deutsch and Martin, 1971).</p>
<ul>
<li>merit: TRUE</li>
<li>optimized by: N/A</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="me-1" class="level2">
<h2 class="anchored" data-anchor-id="me-1">ME</h2>
<p>Measure of effectiveness (McCormick, 1972).</p>
<ul>
<li>merit: TRUE</li>
<li>optimized by: BEA, BEA_TSP</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="moore_stress-1" class="level2">
<h2 class="anchored" data-anchor-id="moore_stress-1">Moore_stress</h2>
<p>Stress criterion (Moore neighborhood) applied to the reordered matrix (Niermann, 2005).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: N/A</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
<section id="neumann_stress-1" class="level2">
<h2 class="anchored" data-anchor-id="neumann_stress-1">Neumann_stress</h2>
<p>Stress criterion (Neumann neighborhood) applied to the reordered matrix (Niermann, 2005).</p>
<ul>
<li>merit: FALSE</li>
<li>optimized by: N/A</li>
</ul>
<p>additional parameters: no parameters</p>
</section>
</section>
<section id="references" class="level1">
<h1>References</h1>
<p>Barnard, S.T., A. Pothen, and H. D. Simon (1993): A Spectral Algorithm for Envelope Reduction of Sparse Matrices. <em>In Proceedings of the 1993 ACM/IEEE Conference on Supercomputing,</em> 493–502. Supercomputing ’93. New York, NY, USA: ACM.</p>
<p>Caraux, G. and S. Pinloche (2005): Permutmatrix: A Graphical Environment to Arrange Gene Expression Profiles in Optimal Linear Order, <em>Bioinformatics,</em> <strong>21</strong>(7), 1280–1281.</p>
<p>Chen, C.-H. (2002): Generalized association plots: Information visualization via iteratively generated correlation matrices, <em>Statistica Sinica,</em> <strong>12</strong>(1), 7–29.</p>
<p>Deutsch, S.B. and J.J. Martin (1971): An ordering algorithm for analysis of data arrays. <em>Operational Research,</em> <strong>19</strong>(6), 1350–1362. </p>
<p>Earle, D. and C.B. Hurley (2015): Advances in Dendrogram Seriation for Application to Visualization. <em>Journal of Computational and Graphical Statistics,</em> <strong>24</strong>(1), 1–25. </p>
<p>Hahsler, M. (2017): An experimental comparison of seriation methods for one-mode two-way data. <em>European Journal of Operational Research,</em> <strong>257</strong>, 133–143. </p>
<p>Hubert, L. and J. Schultz (1976): Quadratic Assignment as a General Data Analysis Strategy. <em>British Journal of Mathematical and Statistical Psychology,</em> <strong>29</strong>(2). Blackwell Publishing Ltd.&nbsp;190–241. </p>
<p>Hubert, L., P. Arabie, and J. Meulman (2001): <em>Combinatorial Data Analysis: Optimization by Dynamic Programming.</em> Society for Industrial Mathematics. </p>
<p>Niermann, S. (2005): Optimizing the Ordering of Tables With Evolutionary Computation, <em>The American Statistician,</em> <strong>59</strong>(1), 41–46. </p>
<p>McCormick, W.T., P.J. Schweitzer and T.W. White (1972): Problem decomposition and data reorganization by a clustering technique, <em>Operations Research,</em> <strong>20</strong>(5), 993-1009. </p>
<p>Robinson, W.S. (1951): A method for chronologically ordering archaeological deposits, <em>American Antiquity,</em> <strong>16</strong>, 293–301. </p>
<p>Tien, Y-J., Yun-Shien Lee, Han-Ming Wu and Chun-Houh Chen (2008): Methods for simultaneously identifying coherent local clusters with smooth global patterns in gene expression profiles, <em>BMC Bioinformatics,</em> <strong>9</strong>(155), 1–16. </p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>